test/app/on_shutdownlib.c

```diff
@@ -0,0 +1,89 @@
+#include <lua.h>
+#include <lauxlib.h>
+#include <lualib.h>
+#include <luaconf.h>
+
+#include <module.h>
+#include "../unit/unit.h"
+
+struct module {
+	/*
+	 * Flag of the module state, May be -1, 0, 1
+	 * -1 means that module thread failed to start
+	 * 0 means that module currently stop
+	 * 1 menas that module currently running
+	 */
+	int is_running;
+	/*
+	 * Module fiber
+	 */
+	struct fiber *fiber;
+};
+
+static struct module module;
+
+/** Shutdown function to test on_shutdown API */
+static int
+on_shutdown_module_bad_func(void *arg)
+{
+	fail_unless(0);
+	return 0;
+}
+
+/** Shutdown function to stop the main fiber of the module */
+static int
+on_shutdown_module_stop_func(void *arg)
+{
+	fprintf(stderr, "stop module fiber\n");
+	fiber_wakeup(module.fiber);
+	fprintf(stderr, "join module fiber\n");
+	fiber_join(module.fiber);
+	fprintf(stderr, "join module fiber finished\n");
+	return 0;
+}
+
+/** Main module thread function. */
+static int
+module_fiber_f(va_list arg)
+{
+
+	__atomic_store_n(&module.is_running, 1, __ATOMIC_SEQ_CST);
+	fiber_yield();
+	fiber_sleep(1);
```

Comment:
Sleep includes yield. Otherwise don't get why we need two yields here

Reply:
In first yield we return control until the on_shutdown_module_stop_func function is called. Sleep needed for test purpose: if sleep time > on_shutdown_wait_time then timeout in trigger_fiber_run expired and tarantool will finish without waiting for the module to finish working

Reply:
Could you please add your comment to the code?